### ISSUE TITLE
Use case for Starlane

### ISSUE THREAD

**Issue by tjsharp1** [FuelLabs/fuel-indexer#59]
Just to describe what we have now, and what we would need from starlane:

The way to get data in seems pretty straightforward, the service could quickly be adapted to send byte blobs to an http endpoint.

From there, we can do all our usual stuff, I think I'd like to understand the route out of the WASM and to the database a little more.

Currently, we expect that the app developer is going to write their own schema in GraphQL, which we then translate into a SQL schema. From the schema, we also generate a bunch of rust types that are used inside the WASM to do queries to the database. This is a glue layer that kind of hides the ugliness of the ffi interface. At the moment, the schemas are flat ( no join tables or anything), but there will very likely be a need for that kind of thing in the future. This right here would be a great advantage if it is already well supported in starlane, since we would not have to do too much work to get those features.

So @uberscott, are these things supported already? Things like pagination, joins, etc? Could I see some examples of how these things are handled? I would want to sweep these details into a proc_macro, so indexer developers don't have to think about these things. Also, if it would help give you a picture of what we're doing, maybe I can show you an example of what we have right now?

**Comment by uberscott** [comment]
First off @tjsharp1 I'm so sorry that I missed this for so long!   Let me endeavor to answer your questions:

1. Starlane does not at this time support pagination or joins... pagination & streaming is scheduled to be developed for a version 1.1+ release (and I'm only just getting close to an initial version 1.0-alpha release). so it may be along ways off.
2. Starlane will be able to support joins simply because it passes that responsibility onto the database itself.    Basically Starlane makes any resource messagable and converts that message into something that the Resource can actually handle using WebAssembly... so there is some WebAssembly code that  you can deploy that may know how to take an Http<Get> message and convert it into something like "SELECT * FROM people,places WHERE people.home_id=place.id".  
3. As for a proc_macro, YES! Starlane has that for capturing the right messages and converting them into code... for example:

> `#[routes]
impl MyMechtron {

  #[route('Http<Get>/plus-one')]
  pub fn plus_one( &self, ctx: InCtx<'_,i64> ) -> i64 {
     ctx.input + 1
  }

}`

I still need to create a proc macro that creates a remote stub that looks something like this:
`pub trait MyMechtronStub {
   fn plus_one( &self, i64 ) -> Result<i64,Err>;
}`

So bottom line, Starlane is not really ready to help you guys yet. 

I actually think the most useful feature that it has you might utilize is the ability to download and execute Wasm in a common Host using Starlane's artifact system.... but to currently use the artifact system the program must be inside Starlane (although enabling it to work externally is a high priority before release of version 1.0) 

Let's keep the dialog going because I have also split the code into about 20 packages that have useful utilities for you without implementling the entire starlane...

**Comment by ra0x3** [comment]
Happy to re-open if we see a rejuvenated use case for this :)

**Comment by uberscott** [comment]
sure
